source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -212,12 +213,16 @@ Network::IoResult TsiSocket::doWrite(Buffer::Instance& buffer, bool end_stream)
 }
 
 void TsiSocket::closeSocket(Network::ConnectionEvent) {
+  ENVOY_CONN_LOG(debug, "TSI: closing socket", callbacks_->connection());
   if (handshaker_) {
     handshaker_.release()->deferredDelete();
   }
 }
 
-void TsiSocket::onConnected() { ASSERT(!handshake_complete_); }
+void TsiSocket::onConnected() {
+  ASSERT(!handshake_complete_);
+  doHandshakeNext();
```

Comment:
I understand that handshaker creation was moved to doHandshakeNext() because you called doHandhsakeNext() here.
But what's the purpose of making a call to doHandshakeNext() earlier at this point?
My worry is that localAddress() may not be initialized so far, can you check that?

Reply:
By doing this it is more efficient to not repeatedly call onHandshakerNext when the transport socket didn't received new data. The first time here we skip the empty check on read_buffer but the rest have. (The first next call for client need to be done when it is empty)

No worries about localAddress, since this is called at same stack that doWrite is called.
https://github.com/envoyproxy/envoy/blob/master/source/common/network/connection_impl.cc#L468

Reply:
Thanks for the explanation! To prevent future changes break this condition, can you add an assert before creating handshaker_ in line:52 to check local_address is initialized?